### PR TITLE
SNOW-1844094: Skip optimization if error is raised during optimization stage

### PR DESCRIPTION
**Pull request by sfc-gh-aalam**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1844094

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.
   - [x] I acknowledge that I have ensured my changes to be thread-safe. Follow the link for more information: [Thread-safe Developer Guidelines](https://github.com/snowflakedb/snowpark-python/blob/main/CONTRIBUTING.md#thread-safe-development)

3. Please describe how your code solves the related issue.

   Wrap optimization in a try-catch to return unoptimized queries when something during optimization stage raises error.

### CODE REVIEW

File: src/snowflake/snowpark/_internal/compiler/plan_compiler.py [hunk 1]
```diff
@@ -77,98 +80,106 @@ def should_start_query_compilation(self) -> bool:
...
-                CompilationStageTelemetryField.TYPE_LARGE_QUERY_BREAKDOWN_OPTIMIZATION_SKIPPED.value: skipped_summary,
-            }
-            # add the extra optimization status
-            summary_value.update(extra_optimization_status)
-            session._conn._telemetry_client.send_query_compilation_summary_telemetry(
-                session_id=session.session_id,
-                plan_uuid=self._plan.uuid,
-                compilation_stage_summary=summary_value,
-            )
-        else:
```
Comment (sfc-gh-helmeleegy) [inline src/snowflake/snowpark/_internal/compiler/plan_compiler.py:166]: Why remove the else branch?

File: src/snowflake/snowpark/_internal/compiler/plan_compiler.py [hunk 2]
```diff
@@ -77,98 +80,106 @@ def should_start_query_compilation(self) -> bool:
...
-
-            extra_optimization_status: Dict[str, Any] = {}
-            # 3. apply each optimizations if needed
-            # CTE optimization
-            cte_start_time = time.time()
-            if session.cte_optimization_enabled:
-                repeated_subquery_eliminator = RepeatedSubqueryElimination(
-                    logical_plans, query_generator
+            try:
+                session = self._plan.session
```
Comment (sfc-gh-helmeleegy) [inline src/snowflake/snowpark/_internal/compiler/plan_compiler.py:92]: Did we have wrong indentation before introducing the `try` block?

File: src/snowflake/snowpark/_internal/compiler/plan_compiler.py (hunk 1 above)
Comment (sfc-gh-helmeleegy) [inline src/snowflake/snowpark/_internal/compiler/plan_compiler.py:166]: Did we have tests exercising this branch? If so, what happens to them?

Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/compiler/plan_compiler.py:166]: we moved the else clause at the top. here is how I did the refactor

old pattern:
```
if should_optimize:
  # apply repeated subquery elimination
  # apply large query breakdown
  queries = optimized_queries
else:
  queries = unoptimized_queries_from_snowflake_plan

return queries
```

new pattern
```
queries =  unoptimized_queries_from_snowflake_plan
if should_optimize:
  try:
    # apply repeated subquery elimination
    # apply large query breakdown
    queries = optimized_queries
  except:
    logger.debug()
```

File: src/snowflake/snowpark/_internal/compiler/plan_compiler.py (hunk 2 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/compiler/plan_compiler.py:92]: the indentation was correct. I would suggest using hide whitespace in diff-view settings. 

Screenshot 2024-12-04 at 9 47 24 AM

File: src/snowflake/snowpark/_internal/compiler/plan_compiler.py [hunk 3]
```diff
@@ -77,98 +80,106 @@ def should_start_query_compilation(self) -> bool:
...
-            )
-        else:
-            final_plan = self._plan
-            queries = {
-                PlanQueryType.QUERIES: final_plan.queries,
-                PlanQueryType.POST_ACTIONS: final_plan.post_actions,
-            }
+            except Exception as e:
+                # if any error occurs during the compilation, we should fall back to the original plan
+                _logger.debug(f"Skipping optimization due to error: {e}")
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/_internal/compiler/plan_compiler.py:181]: Can we also sent a telemetry with error status here?

Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/compiler/plan_compiler.py:181]: done.

**Comment by sfc-gh-aalam** [comment]
https://ci-dev-142.int.snowflakecomputing.com/job/PythonStoredProcBuildPrecommitTest/1495/

**Comment by sfc-gh-aalam** [comment]
https://github.com/snowflakedb/snowpark-python/actions/runs/12169626312

**Review by sfc-gh-yzou — APPROVED** [review]

**Review by sfc-gh-aling — APPROVED** [review]

**Pull request merged**